File: apps/fountain_cli/lib/fountain_cli/keys.ex
defmodule FountainCli.Keys do
  @moduledoc """
  Subcommands for managing Fountain API keys.

      fountain keys list
          Lists API keys: prefix, name, last_used_at.

      fountain keys create <name>
          Creates a new API key and prints the plaintext key once.

      fountain keys revoke <id>
          Revokes the key with the given id (confirmation required).
  """

  alias FountainCli.Api

  def dispatch(["list" | rest]), do: list(rest)
  def dispatch(["create", name | rest]), do: create(name, rest)
  def dispatch(["revoke", id | rest]), do: revoke(id, rest)

  def dispatch(["create" | _]) do
    FountainCli.die("usage: fountain keys create <name>")
  end

  def dispatch(["revoke" | _]) do
    FountainCli.die("usage: fountain keys revoke <id>")
  end

  def dispatch(args) do
    FountainCli.die(
      "unknown keys command: #{Enum.join(args, " ")}\n" <>
        "Available: list, create <name>, revoke <id>"
    )
  end

  # ── list ─────────────────────────────────────────────────

  defp list(_args) do
    case Api.get("/auth/api-keys") do
      {:ok, %{"data" => keys}} when is_list(keys) ->
        print_table(
          ["prefix", "name", "last_used"],
          Enum.map(keys, fn k ->
            [k["key_prefix"], k["name"], k["last_used_at"] || "never"]
          end)
        )

      {:ok, other} ->
        FountainCli.die("unexpected response: #{inspect(other)}")

      {:error, e} ->
        FountainCli.die(inspect(e))
    end
  end

  # ── create ───────────────────────────────────────────────

  defp create(name, _args) do
    case Api.post("/auth/api-keys", %{name: name}) do
      {:ok, %{"data" => %{"key" => key} = k}} ->
        IO.puts("""

        ╭────────────────────────────────────────────────────────────────╮
        │  Save this key — it will not be shown again.                  │
        ╰────────────────────────────────────────────────────────────────╯

        #{key}

        Name:   #{k["name"]}
        Prefix: #{k["key_prefix"]}
        """)

      {:ok, other} ->
        FountainCli.die("unexpected response: #{inspect(other)}")

      {:error, e} ->
        FountainCli.die(inspect(e))
    end
  end

  # ── revoke ──────────────────────────────────────────────

  defp revoke(id, _args) do
    IO.write("Revoke API key #{id}? This cannot be undone. [y/N] ")
    confirm = IO.gets("") |> String.trim() |> String.downcase()

    if confirm in ["y", "yes"] do
      case Api.delete("/auth/api-keys/#{id}") do
        {:ok, _} ->
          IO.puts("Revoked #{id}.")

        {:error, {404, _}} ->
          FountainCli.die("key not found: #{id}")

        {:error, e} ->
          FountainCli.die(inspect(e))
      end
    else
      IO.puts("Aborted.")
    end
  end

  # ── shared helpers ───────────────────────────────────────────

  defp print_table(headers, rows) do
    widths =
      Enum.with_index(headers, fn h, i ->
        [to_string(h) | Enum.map(rows, fn r -> to_string(Enum.at(r, i) || "") end)]
        |> Enum.map(&String.length/1)
        |> Enum.max(fn -> 0 end)
      end)

    IO.puts(
      headers
      |> Enum.with_index()
      |> Enum.map_join("  ", fn {h, i} ->
        String.pad_trailing(to_string(h), Enum.at(widths, i))
      end)
    )

    IO.puts(Enum.map_join(widths, "  ", &String.duplicate("-", &1)))

    for r <- rows do
      IO.puts(
        r
        |> Enum.with_index()
        |> Enum.map_join("  ", fn {v, i} ->
          String.pad_trailing(to_string(v || ""), Enum.at(widths, i))
        end)
      )
    end
  end
end

File: cli/internal/cmd/keys.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"strings"

	"github.com/BinaryBourbon/fountain/cli/internal/api"
	"github.com/BinaryBourbon/fountain/cli/internal/output"
	"github.com/spf13/cobra"
)

func init() {
	keysCmd := &cobra.Command{Use: "keys", Short: "Manage Fountain API keys"}
	listCmd := &cobra.Command{
		Use:   "list",
		Short: "List API keys",
		RunE:  func(cmd *cobra.Command, args []string) error { return keysList(cmd) },
	}
	// Every other list command takes --json; docs/cli.md promises it
	// generally.
	listCmd.Flags().Bool("json", false, "output JSON")
	keysCmd.AddCommand(
		listCmd,
		&cobra.Command{
			Use:   "create <name>",
			Short: "Create a new API key",
			Args:  cobra.ExactArgs(1),
			RunE:  func(cmd *cobra.Command, args []string) error { return keysCreate(args[0]) },
		},
		&cobra.Command{
			Use:   "revoke <id>",
			Short: "Revoke an API key",
			Args:  cobra.ExactArgs(1),
			RunE:  func(cmd *cobra.Command, args []string) error { return keysRevoke(args[0]) },
		},
	)
	rootCmd.AddCommand(keysCmd)
}

// apiKeyCreated mirrors FountainWeb.ApiKeyJSON.created/1
// (apps/fountain/lib/fountain_web/controllers/api_key_json.ex): a flat
// object — no `data` envelope — and the prefix field is `prefix`, not
// `key_prefix`. Decoding the wrong shape here loses the plaintext key
// forever, because this response is the only place it ever appears (#398).
type apiKeyCreated struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Key       string `json:"key"`
	Prefix    string `json:"prefix"`
	CreatedAt string `json:"created_at"`
}

// apiKeySummary mirrors one element of FountainWeb.ApiKeyJSON.index/1, which
// does wrap the list in a `data` envelope.
type apiKeySummary struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	Prefix     string `json:"prefix"`
	CreatedAt  string `json:"created_at"`
	LastUsedAt string `json:"last_used_at"`
}

func keysList(cmd *cobra.Command) error {
	jsonOut, _ := cmd.Flags().GetBool("json")
	c := activeClient()
	var resp struct {
		Data []apiKeySummary `json:"data"`
	}
	if err := c.Get("/auth/api-keys", &resp); err != nil {
		Fatal(err.Error())
	}
	if jsonOut {
		return output.PrintJSON(resp.Data)
	}
	rows := make([][]string, 0, len(resp.Data))
	for _, k := range resp.Data {
		lastUsed := k.LastUsedAt
		if lastUsed == "" {
			lastUsed = "never"
		}
		rows = append(rows, []string{k.Prefix, k.Name, lastUsed})
	}
	output.Table([]string{"prefix", "name", "last_used"}, rows)
	return nil
}

func keysCreate(name string) error {
	c := activeClient()
	var resp apiKeyCreated
	if err := c.Post("/auth/api-keys", map[string]string{"name": name}, &resp); err != nil {
		Fatal(err.Error())
	}
	if resp.Key == "" {
		Fatalf("unexpected response: no key field (id=%q name=%q)", resp.ID, resp.Name)
	}
	fmt.Println()
	fmt.Println("╭────────────────────────────────────────────────────────────────╮")
	fmt.Println("│  Save this key — it will not be shown again.                  │")
	fmt.Println("╰────────────────────────────────────────────────────────────────╯")
	fmt.Println()
	fmt.Println(resp.Key)
	fmt.Println()
	fmt.Printf("Name:   %s\n", resp.Name)
	fmt.Printf("Prefix: %s\n", resp.Prefix)
	fmt.Println()
	return nil
}

func keysRevoke(id string) error {
	fmt.Printf("Revoke API key %s? This cannot be undone. [y/N] ", id)
	r := bufio.NewReader(os.Stdin)
	answer, _ := r.ReadString('\n')
	answer = strings.ToLower(strings.TrimSpace(answer))
	if answer != "y" && answer != "yes" {
		fmt.Println("Aborted.")
		return nil
	}
	c := activeClient()
	if err := c.Delete("/auth/api-keys/"+id, nil); err != nil {
		if api.StatusCode(err) == 404 {
			Fatalf("key not found: %s", id)
		}
		Fatal(err.Error())
	}
	fmt.Printf("Revoked %s.\n", id)
	return nil
}
